scalanet/src/io/iohk/scalanet/peergroup/dynamictls/CustomTlsValidator.scala

```diff
@@ -109,16 +153,48 @@ object CustomTlsValidator {
     * The certificate MUST contain the libp2p Public Key Extension. If this extension is missing,
     * endpoints MUST abort the connection attempt
     *
+    * Endpoints MUST abort the connection attempt if the certificate contains critical extensions
+    * that the endpoint does not understand.
     */
   private def validateCertificateExtension(
       cert: X509Certificate,
       extensionId: String
   ): Either[CertificateError, Array[Byte]] = {
-    val extension = cert.getExtensionValue(extensionId)
-    if (extension == null) {
-      Left(NoCertExtension)
-    } else {
-      Right(extension)
+    Option(cert.getExtensionValue(extensionId)).toRight(NoCertExtension).flatMap { extension =>
+      val criticalExtensions = cert.getCriticalExtensionOIDs
+
+      val containsOnlyOneKnownCriticalExtension = criticalExtensions.size() == 0 ||
+        (criticalExtensions.size() == 1 && criticalExtensions.contains(knownCertificateExtension))
```

Comment:
```suggestion
        (criticalExtensions.size() == 1 && criticalExtensions.contains(SignedKey.extensionIdentifier))
```
Since there's only one `knownCertificateExtension`, I reckon it would be easier if we didn't have to scroll up to the top to see what it is, or if the variable was declared here, so we'd know it's not used for anything other than this check.

Actually since `validateCertificates` passes in `SignedKey.extensionIdentifier` in `extendsionId`, would it make sense to just use `extensionId` in the `contains` check?

Or maybe to break out this check, since it looks like `validateCertificateExtension` is about getting a specific (possibly non-critical) extension, and if we had more than one, it wouldn't make too much sense to validate all known critical extensions every time we read the value of one of the ones we need.